components/sic_image_engine/src/operations/horizontal_gradient.rs

```diff
@@ -0,0 +1,73 @@
+use crate::errors::SicImageEngineError;
+use crate::operations::ImageOperation;
+use crate::wrapper::gradient_input::GradientInput;
+use rayon::iter::{IntoParallelRefMutIterator, ParallelIterator};
+use sic_core::{
+    image::{self, DynamicImage, Frame, Pixel},
+    SicImage,
+};
+
+#[derive(Clone, Debug, Copy, PartialEq)]
+pub struct HorizontalGradient {
+    pub colors: GradientInput,
+}
+
+impl HorizontalGradient {
+    pub fn new(colors: GradientInput) -> Self {
+        Self { colors }
+    }
+}
+
+impl ImageOperation for HorizontalGradient {
+    fn apply_operation(&self, image: &mut SicImage) -> Result<(), SicImageEngineError> {
+        match image {
+            SicImage::Static(image) => gradient_static_image(image, self.colors),
+            SicImage::Animated(image) => gradient_animated_image(image.frames_mut(), self.colors),
+        }
+
+        Ok(())
+    }
+}
+
+fn gradient_static_image(img: &mut DynamicImage, input: GradientInput) {
+    let (left_color, right_color) = &input.colors();
+    let (width, height) = img.to_rgba8().dimensions();
+    let mut gradient_dynamic = DynamicImage::new_rgba8(width, height);
+    image::imageops::horizontal_gradient(&mut gradient_dynamic, left_color, right_color);
+    blend_static_image(img, gradient_dynamic);
+}
+
+fn gradient_animated_image(frames: &mut [image::Frame], input: GradientInput) {
+    let (left_color, right_color) = &input.colors();
+    let (width, height) = DynamicImage::ImageRgba8(frames[0].buffer().clone())
+        .to_rgba8()
+        .dimensions();
+    let mut gradient_dynamic = DynamicImage::new_rgba8(width, height);
+    image::imageops::horizontal_gradient(&mut gradient_dynamic, left_color, right_color);
+    let gradient_buffer = gradient_dynamic.to_rgba8();
+
+    frames.par_iter_mut().for_each(|frame| {
+        blend_frame(frame, &gradient_buffer);
+    });
+}
+
+fn blend_frame(frame: &mut Frame, layer: &image::ImageBuffer<image::Rgba<u8>, Vec<u8>>) {
+    let mut gradient_iter = layer.pixels();
+    let mut blended_buffer = DynamicImage::ImageRgba8(frame.buffer().clone()).into_rgba8();
+
+    blended_buffer
+        .pixels_mut()
+        .for_each(|pixel| pixel.blend(gradient_iter.next().unwrap()));
+    *frame = Frame::new(blended_buffer);
+}
```

Comment:
Similar to static image (see also below :)), we can iterate over the two pixel buffers using `Iterator::zip`, and we can mutate the pixels directly:

```rust
fn blend_frame(frame: &mut Frame, layer: &RgbaImage) {
    *frame
        .buffer_mut()
        .pixels_mut()
        .zip(layer.pixels())
        .for_each(|(source_pixel, gradient_pixel)| source_pixel.blend(gradient_pixel));
}
```

Compared to the static image, we don't need to convert the frame (`blended_buffer` below) to an `ImageBuffer` with pixel type `Rgba`. This is because in the current version of the image crate, each frame [already consists](https://docs.rs/image/0.23.14/src/image/animation.rs.html#46) of an `RgbaImage` buffer.

Otherwise we do pretty much the same thing :smiley:: iterate over all pixels in the buffer, and blend them with the pixels from the gradient side buffer.

One last thing of note: instead of creating a new frame, we should use re-use the existing one (we do that with the `.buffer_mut()` above) and directly access the pixel buffer instead (the `.pixels_mut()`).

The reason for this is that the frame consists of more than the `ImageBuffer`; of note is the frame `delay`. If we create a new frame, the delay will be the default value instead of the value as imported.

If we take as input value:
![loop](https://user-images.githubusercontent.com/5955761/110901263-5854ea80-8304-11eb-835d-b88cd2f24ed3.gif)

And then apply the gradient while creating new frames we get:

![was](https://user-images.githubusercontent.com/5955761/110901527-c00b3580-8304-11eb-9d5e-cc7d2a4ebddf.gif)


If we re-use the frames as I did in the example above, we instead get:

![expected](https://user-images.githubusercontent.com/5955761/110901376-8a664c80-8304-11eb-9484-f0c1cf64cbcc.gif)

Reply:
You may also note the artefacts on some pixels around some of the numbers. I'm not entirely sure yet why that happens. Possibly this has to do with the type of blending, but this is no blocker for me.

Reply:
ah i see i made some expensive calls 🤦🏾  , thanks for the helpful review i learned a lot, i will try to change


> You may also note the artefacts on some pixels around some of the numbers. I'm not entirely sure yet why that happens. Possibly this has to do with the type of blending, but this is no blocker for me.

i will try to investigate about that too after i finish